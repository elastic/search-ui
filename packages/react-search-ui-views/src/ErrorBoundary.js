import PropTypes from "prop-types";
import React from "react";

import { appendClassName } from "./view-helpers";

function ErrorBoundary({ children, className, error, ...rest }) {
  if (error) {
    return (
      <div className={appendClassName("sui-search-error", className)} {...rest}>
        {error}
      </div>
    );
  }

  return children;
}

ErrorBoundary.propTypes = {
  children: PropTypes.node.isRequired,
  error: PropTypes.string.isRequired,
  className: PropTypes.string
};

export default ErrorBoundary;
