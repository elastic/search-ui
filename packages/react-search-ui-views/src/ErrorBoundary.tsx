import React from "react";

import { appendClassName } from "./view-helpers";
import type { SearchContextState } from "@elastic/search-ui";
import { BaseContainerProps } from "./types";

export type ErrorBoundaryContainerContext = Pick<SearchContextState, "error">;

export type ErrorBoundaryViewProps = BaseContainerProps &
  ErrorBoundaryContainerContext;

function ErrorBoundary({
  children,
  className,
  error,
  ...rest
}: ErrorBoundaryViewProps & React.HTMLAttributes<HTMLDivElement>): JSX.Element {
  if (error) {
    return (
      <div className={appendClassName("sui-search-error", className)} {...rest}>
        {error}
      </div>
    );
  }

  return children as JSX.Element;
}

export default ErrorBoundary;
