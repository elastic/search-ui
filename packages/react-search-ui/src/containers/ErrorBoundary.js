import PropTypes from "prop-types";
import { Component } from "react";
import { withSearch } from "..";
import { ErrorBoundary } from "@elastic/react-search-ui-views";

export class ErrorBoundaryContainer extends Component {
  static propTypes = {
    // Props
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    view: PropTypes.func,
    // State
    error: PropTypes.string.isRequired
  };

  render() {
    const { children, className, error, view, ...rest } = this.props;

    const View = view || ErrorBoundary;

    return View({
      className,
      children,
      error,
      ...rest
    });
  }
}

export default withSearch(({ error }) => ({ error }))(ErrorBoundaryContainer);
