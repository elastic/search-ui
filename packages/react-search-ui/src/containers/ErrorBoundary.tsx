import React, { Component } from "react";
import { withSearch } from "..";
import {
  ErrorBoundary,
  ErrorBoundaryContainerContext,
  ErrorBoundaryViewProps
} from "@elastic/react-search-ui-views";
import { BaseContainerProps } from "../types";

type ErrorBoundaryContainerProps = BaseContainerProps &
  ErrorBoundaryContainerContext & {
    view?: React.ComponentType<ErrorBoundaryViewProps>;
  };

export class ErrorBoundaryContainer extends Component<ErrorBoundaryContainerProps> {
  render() {
    const { children, className, error, view, ...rest } = this.props;

    const View = view || ErrorBoundary;

    const viewProps = {
      className,
      children,
      error,
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<
  ErrorBoundaryContainerProps,
  ErrorBoundaryContainerContext
>(({ error }) => ({ error }))(ErrorBoundaryContainer);
