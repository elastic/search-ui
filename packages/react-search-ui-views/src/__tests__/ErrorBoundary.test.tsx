import React from "react";
import ErrorBoundary from "../ErrorBoundary";
import { shallow } from "enzyme";

const params = {
  children: <div>Child</div>,
  error: "I am an error"
};

it("renders an error when there is an error", () => {
  const wrapper = shallow(<ErrorBoundary {...params} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders children when there is no error", () => {
  const wrapper = shallow(
    <ErrorBoundary
      {...{
        ...params,
        error: ""
      }}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const wrapper = shallow(
    <ErrorBoundary className={customClassName} {...params} />
  );
  const { className } = wrapper.props();
  expect(className).toEqual("sui-search-error test-class");
});
