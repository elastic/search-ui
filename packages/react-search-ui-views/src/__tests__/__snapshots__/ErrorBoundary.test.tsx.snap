// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an error when there is an error 1`] = `
<div
  className="sui-search-error"
>
  I am an error
</div>
`;

exports[`renders children when there is no error 1`] = `
<div>
  Child
</div>
`;
